lighthouse-core/computed/critical-request-chains.js

```diff
@@ -66,6 +67,53 @@ class CriticalRequestChains {
     return ['VeryHigh', 'High', 'Medium'].includes(request.priority);
   }
 
+  /**
+   * Create a tree of critical requests.
+   * @param {LH.Artifacts.NetworkRequest} mainResource
+   * @param {LH.Gatherer.Simulation.GraphNode} graph
+   * @return {LH.Artifacts.CriticalRequestNode}
+   */
+  static extractChainUsingLantern(mainResource, graph) {
+    /** @type {LH.Artifacts.CriticalRequestNode} */
+    const rootNode = {};
+
+    /**
+     * @param {LH.Gatherer.Simulation.GraphNetworkNode[]} path
+     */
+    function addChain(path) {
+      let currentNode = rootNode;
+
+      for (const node of path) {
+        if (!currentNode[node.id]) {
+          currentNode[node.id] = {
+            request: node.record,
+            children: {},
+          };
+        }
+
+        currentNode = currentNode[node.id].children;
+      }
+    }
+
+    graph.traverse((node, traversalPath) => {
+      if (node.type !== 'network') return;
+      if (!CriticalRequestChains.isCritical(node.record, mainResource)) return;
+
+      const networkPath = traversalPath
```

Comment:
~~Can you add a comment here that the traversalPath is not necessarily deterministic because it uses dependencies instead of initiators (so which one it picks might not be the same every time)?~~

EDIT: Now that I'm thinking about it more, I think this means we're going to be in a bit of trouble with assets that *also* have a dependency on the root document because they'll have their important traversalPath skipped in this traversal (traverse is BFS and so `traversalPath` is the *shortest* path each request, and we kinda want the opposite here). 

Some options: 
- Use initiators/`getDependencies` on each node here instead of the `traversalPath` and do the ancestor bookkeeping similar to existing `extractChain`
- Write a version of `.traverse` that reports the longest paths instead of the shortest ones and leave this as-is. I *think* you could actually accomplish this using the `getNextNodes` argument of traverse where you only allow traversal to a node once you've seen *all* of its dependencies instead of just one of them.